### ISSUE TITLE
Validator Usage Problem

### ISSUE THREAD

**Issue by M3ssman** [OCR-D/core#545]
Hello OCR-Ders,

unfortunately, I can't figure out how to use the XsdValidator. 
I'd like to utilize this for Schema-checks.

```
 validator = XsdValidator('http://www.loc.gov/standards/mets/version18/mets.xsd')

=> XML schema not bundled with OCR-D: http://www.loc.gov/standards/mets/version18/mets.xsd
```

From the parameter doc I was expecting something more in terms of an external URL.

**Comment by kba** [comment]
> Hello OCR-Ders,
> 
> unfortunately, I can't figure out how to use the XsdValidator.
> I'd like to utilize this for Schema-checks.
> 
> ```
>  validator = XsdValidator('http://www.loc.gov/standards/mets/version18/mets.xsd')
> 
> => XML schema not bundled with OCR-D: http://www.loc.gov/standards/mets/version18/mets.xsd
> ```

From tests/validator/test_xsd_validator.py:

```python
from ocrd_validators import XsdValidator, XsdMetsValidator, XsdPageValidator
from ocrd_validators.constants import XSD_METS_URL, XSD_PAGE_URL

class TestXsdValidator(TestCase):

    def test_validate_simple_protected_str(self):
        val = XsdValidator(XSD_METS_URL)
        report = val._validate(self.ws.mets.to_xml())
        self.assertTrue(report.is_valid)

    def test_validate_simple_protected_doc(self):
        val = XsdValidator(XSD_METS_URL)
        report = val._validate(self.ws.mets._tree)
        self.assertTrue(report.is_valid)

    def test_validate_simple_static_doc(self):
        report = XsdValidator.validate(XSD_METS_URL, self.ws.mets._tree)
        self.assertTrue(report.is_valid)
```


    def test_mets_empty(self):
        with TemporaryDirectory() as tempdir:
            mets_path = Path(tempdir, 'mets.xml')
            mets_path.write_bytes(METS_XML_EMPTY)
            report = XsdMetsValidator.validate(mets_path)
            self.assertEqual(len(report.errors), 2)
            self.assertEqual(report.errors[0],
                    "Line 3: Element '{http://www.loc.gov/METS/}metsHdr', attribute 'CREATEDATE': '{{ NOW }}' is not a valid value of the atomic type 'xs:dateTime'.")
            self.assertEqual(report.errors[1],
                    "Line 18: Element '{http://www.loc.gov/METS/}fileSec': Missing child element(s). Expected is ( {http://www.loc.gov/METS/}fileGrp ).")
            self.assertFalse(report.is_valid)
```

So, the easiest way to use METS XSD Validation is 

```python
from ocrd_validators import XsdMetsValidator

XsdMetsValidator.validate(file_name_or_etree_doc)
```

Your usage also works but you have to use the version-independent METS XSD URL (also a constant: `from ocrd_validators.constants import XSD_METS_URL`):

```python
XsdValidator('https://www.loc.gov/standards/mets/mets.xsd')._validate(...)
```

I recommend using the `validate` class method though.

> 
> From the parameter doc I was expecting something more in terms of an external URL.

You mean the docstrings? I'm happy to clarify usage or merge a PR for better docs.. We're not totally flexible with support for arbitrary XSD URL because we want to use bundled XSD to reduce networking issues, esp. for imported schemas.

**Comment by M3ssman** [comment]
@kba Thanks for clarifications!

```
SCHEMA_URL = 'https://www.loc.gov/standards/mets/mets.xsd'
doc_to_check = lxtree.parse(meta_written)
validation_report = XsdValidator.validate(SCHEMA_URL, doc_to_check)
assert not validation_report.errors
assert validation_report.is_valid
```
This works fine!

... but it's not happy with the XSD constant import (V2.10):
```
test/test_preprocess_vlmetada.py:17: in <module>
    from ocrd.constants import XSD_METS_URL
E   ImportError: cannot import name 'XSD_METS_URL'
```
Whereas this is okay: 
```
from ocrd import XsdValidator
```

And yes, I will be happy to assist with the docs but I need to understand first ;)
Was this intended only for internal usage?

**Comment by kba** [comment]
> ... but it's not happy with the XSD constant import (V2.10):
> 
> ```
> test/test_preprocess_vlmetada.py:17: in <module>
>     from ocrd.constants import XSD_METS_URL
> E   ImportError: cannot import name 'XSD_METS_URL'
> ```

XSD_METS_URL is defined in `ocrd_validators.constants` not `ocrd.constants`. We could reimport them from `ocrd.constants` if that makes life easier for you.
 
> And yes, I will be happy to assist with the docs but I need to understand first ;)
> Was this intended only for internal usage?

Great, thanks. Was not intended for internal usage only but I thought the `ocrd workspace validate` CLI would be the go-to solution to make use of OCR-D's validation.

**Comment by M3ssman** [comment]
Thanks, with your help Import works!

It looked like a good way to validate written/updated metadata against a schema.

**Comment by kba** [comment]
> (V2.10)

Consider updating to 2.12.3, there have been a lot of fixes/improvements since 2.10.0
